### PR TITLE
Implement Boolean Implementation

### LINKED ISSUE #12423: Codegen:Boolean Implementation
```
>>> from sympy import *
>>> from sympy.abc import x,y,z
>>> from sympy.utilities.codegen import codegen
>>> [(c,d),(e,f)]=codegen(("f",x&y),"C",header=False,empty=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python27\lib\site-packages\sympy\utilities\codegen.py", line 1744, in
 codegen
    global_vars))
  File "C:\Python27\lib\site-packages\sympy\utilities\codegen.py", line 585, in
routine
    return_val.append(Result(expr))
  File "C:\Python27\lib\site-packages\sympy\utilities\codegen.py", line 475, in
__init__
    raise TypeError("The first argument must be a sympy expression.")
TypeError: The first argument must be a sympy expression.
```
I think this is a bug in sympy it should return
```
#include "f.h"
#include <math.h>
double f(double x, double y) {
   double f_result;
   f_result = x && y;
   return f_result;
}
```

### PR DESCRIPTION
**Pull request by SatyaPrakashDwibedi**
Implementation Boolean `codegeneration` ability by adding a check in `Result` class for Boolean-type expression
Fixes #12423 
review the changes done and suggest where to add the corresponding test case in `test_codegen.py`

### CODE REVIEW

**Comment by SatyaPrakashDwibedi** [comment]
@moorepants review the code and suggest changes

**Comment by asmeurer** [comment]
I would just change Expr to Basic. There's no reason to expect code generatable expressions to only be Expr classes, with those exceptions. You could also remove MatrixExpr (subclasses from Basic).

**Comment by SatyaPrakashDwibedi** [comment]
@asmeurer check this

**Comment by asmeurer** [comment]
MatrixBase should be left in. It doesn't subclass from Basic.

**Comment by SatyaPrakashDwibedi** [comment]
@asmeurer Thanks, I have added MatrixBase but I have a doubt here Why isn't MatrixBase is a subclass of Basic whereas MatrixExpr is?

File: sympy/utilities/codegen.py [hunk 1]
```diff
@@ -470,7 +470,8 @@ def __init__(self, expr, name=None, result_var=None, datatype=None,
             Controls the precision of floating point constants.
 
         """
-        if not isinstance(expr, (Expr, MatrixBase, MatrixExpr)):
+# Basic because it is the base class for all types of expressions
```
Comment (asmeurer) [inline sympy/utilities/codegen.py]: Indent this line

**Comment by siefkenj** [comment]
@SatyaPrakashDwibedi `Matrix` is mutable but everything that subclasses from `Basic` must be immutable.

**Comment by SatyaPrakashDwibedi** [comment]
@siefkenj I don't think that matters here, here we are only performing a check that if the input expression is a valid Sympy expression or not.
Reg. the issue #12423, `codegen` did not detect `x&y` as valid expression.

**Comment by siefkenj** [comment]
@SatyaPrakashDwibedi I was answering your question about why `MatrixBase` doesn't inherit from `Basic`.

**Comment by SatyaPrakashDwibedi** [comment]
@siefkenj Ok Thanks 👍

**Pull request merged**